src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs

```diff
@@ -33,6 +34,14 @@ public static async Task PostFaultAsync(Exception e, string callerClassName, [Ca
 
             var fault = new FaultEvent($"{VSTelemetrySession.VSEventNamePrefix}Fault", description, FaultSeverity.General, e, gatherEventDetails: null);
             fault.Properties[$"{VSTelemetrySession.VSPropertyNamePrefix}Fault.Caller"] = caller;
+            if (extraProperties != null)
+            {
+                foreach (var kvp in extraProperties)
+                {
+                    fault.Properties[VSTelemetrySession.VSEventNamePrefix + kvp.Key] = kvp.Value;
```

Comment:
I don't know if the C# compiler is better now, but initially at least interpolated strings  (`$"... {something}"` was just syntactic sugar for `string.Format`, and the book "Writing High Performance .NET Code" discusses how `string.Format` harms performance since the runtime has to parse the format string, and build up the final string. Their recommendation was to use string concatenation instead. Hence I changed line 36 to do that instead.

Reply:
Didn't realize string.Format had that perf hit. I think StringBuilder is the most perf and memory efficient, right?
I don't really prefer reading the concatenation, especially inline an indexer like that.

Reply:
c#10 (or maybe later) might have [const interpolated strings](https://github.com/dotnet/csharplang/issues/2951), so in the future this may no longer be necessary. But `string.Concat` can produce const strings determined at compile time, rather than runtime. And at runtime, `string.Concat` (which the `+` operator is syntactic sugar for) remains to be by far the fastest execution at runtime as well.

`string.Concat` only takes strings as input, anything that is not a string is implicitly run though `.ToString()` before being passed to `string.Concat`, meaning that `string.Concat` can use `args.Sum(a => a.Length)` to determine the exact length of the final string, do a single memory allocation of the required length, then copy the input strings one by one.

Using a `StringBuilder` is the next best. It has a small allocation for the `StringBuilder` object iteself, and it allocates at least one buffer. As you call `Append`, it copies your string into the buffer, and if the buffer gets full, it needs to allocate more buffers. Finally, when you call `StringBuilder.ToString()`, it can do the same work as `string.Concat`, which as previously stated is an additional allocation of the required length. Hence, we can see that `StringBuilder` cannot possibly be more efficient than `string.Concat`, since it does everything that `string.Concat` does, plus more.

`string.Format` (which is what interpolated strings are, and also what StringBuilder's `AppendFormat` method is) is the least efficient. Internally, it creates a string builder (except in the StringBuilder.AppendFormat case), since it can't figure out what the final string length is going to be from the inputs alone. Therefore it can't possibly be faster than `StringBuilder`, since it uses a `StringBuilder`. As previously mentioned, Format also needs to parse the input string to find the `{x}` placeholders, copy the static parts, then parse the placeholder to see if there are formatting options, then call `ToString` on the relevant object with the discovered formatting options. `string.Format` and `StringBuilder.AppendFormat` are therefore by far the slowest at runtime. My memory is that the Writing High Performance .NET Code book claimed that using `StringBuilder` with a series of `Append` calls is faster than `string.Format`.

So in summary, `constStr1 + constStr2` will be evaluated at compile time, so runtime does nothing more than load a const string. This should become available for interpolated const strings in the future, but is not yet available. If the strings are not const at compile time, string concatenation remains the fastest, so if you want to hyper-optimize allocations, this is the best bet. It's always best if all strings you need to concat are in a single call, but even if you use `string.Concat` a small number of times to generate a single string, it can still be faster than using a `StringBuilder`. Using (non-const) interpolated strings, or `string.Format` basically means you don't care about performance. Having said that, while we don't want our tools to be slow, client tools don't have the same concerns or cost savings that web services like Bing or Office web applications have.

Reply:
> string.Concat only takes strings as input

Strings that are already allocated. I thought the whole memory consumption (not perf) problem with concat was in `string1 + string2`, you allocate 2 strings, then you allocate a result string and garbage collect 1 and 2. StringBuilder's buffering avoided that. 
Anyway, thanks for the writeup and book recommendation!